### PR TITLE
[Merged by Bors] - feat: port Data.Polynomial.Splits

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Polynomial/Splits.lean [hunk 1]
```diff
@@ -185,11 +193,11 @@ theorem roots_ne_zero_of_splits' {f : K[X]} (hs : Splits i f) (hf0 : natDegree (
   let ⟨x, hx⟩ := exists_root_of_splits' i hs fun h => hf0 <| natDegree_eq_of_degree_eq_some h
   fun h => by
   rw [← eval_map] at hx
-  cases h.subst ((mem_roots _).2 hx)
-  exact ne_zero_of_nat_degree_gt (Nat.pos_of_ne_zero hf0)
+  have : f.map i ≠ 0 := by intro; simp_all
+  cases h.subst ((mem_roots this).2 hx)
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Splits.lean]: What goes wrong with the original?

File: Mathlib/Data/Polynomial/Splits.lean [hunk 2]
```diff
@@ -313,65 +326,69 @@ theorem map_rootOfSplits {f : K[X]} (hf : f.Splits i) (hfd) :
 #align polynomial.map_root_of_splits Polynomial.map_rootOfSplits
 
 theorem natDegree_eq_card_roots {p : K[X]} {i : K →+* L} (hsplit : Splits i p) :
-    p.natDegree = (p.map i).roots.card :=
+    p.natDegree = Multiset.card (p.map i).roots :=
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Splits.lean]: Dot notation really doesn't work here? That's very strange.

File: Mathlib/Data/Polynomial/Splits.lean [hunk 3]
```diff
@@ -464,20 +485,18 @@ theorem prod_roots_eq_coeff_zero_of_monic_of_split {P : K[X]} (hmo : P.Monic)
     congr
     ext
     rw [neg_eq_neg_one_mul]
-  rw [Multiset.prod_map_mul, Multiset.map_const, Multiset.prod_replicate, Multiset.map_id',
-    splits_iff_card_roots.1 hP]
+  simp only [splits_iff_card_roots.1 hP, neg_mul, one_mul, Multiset.prod_map_neg]
```
Comment (eric-wieser) [inline Mathlib/Data/Polynomial/Splits.lean]: Do you know which step of the `rw` failed?

**Review by eric-wieser — COMMENTED** [review]
An initial skim. Obviously the `set_option`s are a concern, but the approach of getting everything else working first, (by raising the timeouts) and then fixing the timeouts later is a sensible one.

File: Mathlib/Data/Polynomial/Splits.lean (hunk 2 above)
Comment (Parcly-Taxel) [reply Mathlib/Data/Polynomial/Splits.lean]: It really doesn't work everywhere it appears, with the error message
```
invalid field notation, function 'Multiset.card' does not have argument with type (Multiset ...) that can be used, it must be explicit or implicit with a unique name
```

File: Mathlib/Data/Polynomial/Splits.lean (hunk 1 above)
Comment (Parcly-Taxel) [reply Mathlib/Data/Polynomial/Splits.lean]: Lean can't synthesise the term `map i f ≠ 0`, which would fill the underscore.

File: Mathlib/Data/Polynomial/Splits.lean (hunk 3 above)
Comment (Parcly-Taxel) [reply Mathlib/Data/Polynomial/Splits.lean]: The rewrite fails at `Multiset.map_const`. Like the third remaining error in the file at `image_rootSet`, visually the pattern does match but Lean is not getting it.

**Comment by Ruben-VandeVelde** [comment]
This fixes the remaining errors:
```diff
--- Mathlib/Data/Polynomial/Splits.lean
+++ Mathlib/Data/Polynomial/Splits.lean
@@ -208,7 +208,7 @@ theorem map_rootOfSplits' {f : K[X]} (hf : f.Splits i) (hfd) :
   Classical.choose_spec <| exists_root_of_splits' i hf hfd
 #align polynomial.map_root_of_splits' Polynomial.map_rootOfSplits'
 
-attribute [-instance] Ring.toNonAssocRing in -- Porting note: gets around lean4#2074
+set_option synthInstance.etaExperiment true in -- Porting note: gets around lean4#2074
 theorem natDegree_eq_card_roots' {p : K[X]} {i : K →+* L} (hsplit : Splits i p) :
     (p.map i).natDegree = Multiset.card (p.map i).roots := by
   by_cases hp : p.map i = 0
@@ -220,7 +220,7 @@ theorem natDegree_eq_card_roots' {p : K[X]} {i : K →+* L} (hsplit : Splits i p
   rw [← hd, add_right_eq_self]
   by_contra h
   have h' : (map (RingHom.id L) q).natDegree ≠ 0 := by simp [h]
-  have := roots_ne_zero_of_splits' (RingHom.id L) (splits_of_splits_mul' _ _ hsplit).2 h'
+  have := roots_ne_zero_of_splits' (RingHom.id L) (splits_of_splits_mul' _ ?_ hsplit).2 h'
   · rw [map_id] at this
     exact this hr
   · rw [map_id]
@@ -345,9 +345,10 @@ theorem roots_map {f : K[X]} (hf : f.Splits <| RingHom.id K) : (f.map i).roots =
 set_option synthInstance.etaExperiment true in -- Porting note: gets around lean4#2074
 theorem image_rootSet [Algebra F K] [Algebra F L] {p : F[X]} (h : p.Splits (algebraMap F K))
     (f : K →ₐ[F] L) : f '' p.rootSet K = p.rootSet L := by
-  classical rw [rootSet, ← Finset.coe_image, ← Multiset.toFinset_map, ← f.coe_toRingHom, ←
-      roots_map (↑f) ((splits_id_iff_splits (algebraMap F K)).mpr h), map_map, f.comp_algebra_map, ←
-      root_set]
+  classical
+    rw [rootSet, ← Finset.coe_image, ← Multiset.toFinset_map, ← f.coe_toRingHom,
+      ← roots_map _ ((splits_id_iff_splits (algebraMap F K)).mpr h), map_map, f.comp_algebraMap,
+      ← rootSet]
 #align polynomial.image_root_set Polynomial.image_rootSet
 
 set_option synthInstance.etaExperiment true in -- Porting note: gets around lean4#2074
@@ -444,9 +445,10 @@ theorem splits_iff_exists_multiset {f : K[X]} :
 #align polynomial.splits_iff_exists_multiset Polynomial.splits_iff_exists_multiset
 
 theorem splits_comp_of_splits (j : L →+* F) {f : K[X]} (h : Splits i f) : Splits (j.comp i) f := by
-  change i with (RingHom.id _).comp i at h
+  -- Porting note: was
+  -- change i with (RingHom.id _).comp i at h
   rw [← splits_map_iff]
-  rw [← splits_map_iff i] at h
+  rw [← RingHom.id_comp i, ← splits_map_iff i] at h
   exact splits_of_splits_id _ h
 #align polynomial.splits_comp_of_splits Polynomial.splits_comp_of_splits
```

File: Mathlib/Data/Polynomial/Splits.lean (hunk 2 above)
Comment (ChrisHughes24) [reply Mathlib/Data/Polynomial/Splits.lean]: > Dot notation really doesn't work here? That's very strange.

It's because `Multiset.card` is an `AddHom`

**Pull request closed without merging**